node/pkg/query/request.go

```diff
@@ -0,0 +1,470 @@
+	CallData []*EthCallData
+}
+
+// EthCallData specifies the parameters to a single EVM eth_call request.
+type EthCallData struct {
+	// To specifies the contract address to be queried.
+	To []byte
+
+	// Data is the ABI encoded parameters to the query.
+	Data []byte
+}
+
+const EvmContractAddressLength = 20
+
+// PerChainQueryInternal is an internal representation of a query request that is passed to the watcher.
+type PerChainQueryInternal struct {
+	RequestID  string
+	RequestIdx int
+	Request    *PerChainQueryRequest
+}
+
+// QueryRequestDigest returns the query signing prefix based on the environment.
+func QueryRequestDigest(env common.Environment, b []byte) ethCommon.Hash {
+	// TODO: should this use a different standard of signing messages, like https://eips.ethereum.org/EIPS/eip-712
+	var queryRequestPrefix []byte
+	if env == common.MainNet {
+		queryRequestPrefix = []byte("mainnet_query_request_000000000000|")
+	} else if env == common.TestNet {
+		queryRequestPrefix = []byte("testnet_query_request_000000000000|")
+	} else {
+		queryRequestPrefix = []byte("devnet_query_request_0000000000000|")
+	}
+
+	return ethCrypto.Keccak256Hash(append(queryRequestPrefix, b...))
+}
+
+// PostSignedQueryRequest posts a signed query request to the specified channel.
+func PostSignedQueryRequest(signedQueryReqSendC chan<- *gossipv1.SignedQueryRequest, req *gossipv1.SignedQueryRequest) error {
+	select {
+	case signedQueryReqSendC <- req:
+		return nil
+	default:
+		return common.ErrChanFull
+	}
+}
+
+//
+// Implementation of QueryRequest.
+//
+
+// Marshal serializes the binary representation of a query request.
+// This method calls Validate() and relies on it to range checks lengths, etc.
+func (queryRequest *QueryRequest) Marshal() ([]byte, error) {
+	if err := queryRequest.Validate(); err != nil {
+		return nil, err
+	}
+
+	buf := new(bytes.Buffer)
+
+	vaa.MustWrite(buf, binary.BigEndian, uint8(1))           // version
+	vaa.MustWrite(buf, binary.BigEndian, queryRequest.Nonce) // uint32
+
+	vaa.MustWrite(buf, binary.BigEndian, uint8(len(queryRequest.PerChainQueries)))
+	for _, perChainQuery := range queryRequest.PerChainQueries {
+		pcqBuf, err := perChainQuery.Marshal()
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal per chain query: %w", err)
+		}
+		buf.Write(pcqBuf)
+	}
+
+	return buf.Bytes(), nil
+}
+
+// Unmarshal deserializes the binary representation of a query request from a byte array
+func (queryRequest *QueryRequest) Unmarshal(data []byte) error {
+	reader := bytes.NewReader(data[:])
+	return queryRequest.UnmarshalFromReader(reader)
+}
+
+// UnmarshalFromReader deserializes the binary representation of a query request from an existing reader
+func (queryRequest *QueryRequest) UnmarshalFromReader(reader *bytes.Reader) error {
+	var version uint8
+	if err := binary.Read(reader, binary.BigEndian, &version); err != nil {
+		return fmt.Errorf("failed to read message version: %w", err)
+	}
+
+	if version != 1 {
```

Comment:
Please make a constant for this and in the marshal function above. It is too easy to make mistakes with magic numbers.

Reply:
Should be handled in #3423